syntax = "proto3";

package mud;

message player {
	string name = 1;
	string passwordHash = 2;
	int64 id = 3;
	repeated int64 idCharacters = 4;
}

message player_book {
	repeated player players = 1;
}

message direction {
	enum direction_enum {
		NORTH = 0;
		SOUTH = 1;
		EAST = 2;
		WEST = 3;
	}
	direction_enum value = 1;
}

message attribute {
	enum attribute_name {
		LIFE = 0;
		STRENGTH = 1;
		AGILITY = 2;
		INTELLIGENCE = 3;
	}
	
	attribute_name name = 1;
	int32 value = 2;
	int32 regen = 3;
}

message character_state {
	enum character_state_enum{
		NONE = 0;
		WALKING = 1;
		COMBAT = 2;
	}
	character_state_enum value = 1;
}
	
message character {
	string name = 1;
	int64 id = 2;
	int64 tile_id = 3;
	direction facing = 4;
	repeated attribute attributes = 5;
	character_state state = 6;
}

message enemy {
	string name = 1;
	int64 id = 2;
	int64 tile_id = 3;
	direction facing = 4;
	repeated attribute attributes = 5;
	character_state state = 6;
}

message character_book {
	repeated character characters = 1;
}

message enemy_book {
	repeated enemy enemies = 1;
}



message neighbour {
	direction neighbour_direction = 1;
	int64 neighbour_tile_id = 2;
}



message tile {
	enum tile_type {
		EMPTY = 0;
		WALL = 1;
		TREE = 2;
		PORTAL = 3;
	}

	enum resident_type {
		NOBODY = 0;
		CHARACTER = 1;
		ENEMY = 2;
	}
	
	int64 id = 1;
	tile_type type = 2;
	repeated neighbour neighbours = 3;
	resident_type occupant_type = 4;
	int64 occupant_id = 5;
	string mood = 6;
}

message tile_book {
	repeated tile tiles = 1;
}

message token_in {}

message token_out {
	int64 token = 1;
}

message login_in {
	string name = 1;
	string hash = 2;
	int64 token = 3;
}

message login_out {
	enum login_state_enum{
		OK = 0;
		FAILED = 1;
	}
	repeated character characters = 1;
	login_state_enum login_state = 2;
}

message select_character_in {
	enum action_type{
		SELECT = 0;
		CREATE = 1;
	}
	string name = 1;
	int64 token = 2;
	action_type action = 3;
}

message select_character_out {
	enum result_enum{
		OK = 0;
		NOT_FOUND = 1;
		NAME_ALREADY_EXIST = 2;
		ERROR = 3;
	}
	character selected_character = 1;
	result_enum result = 2;
}

service game_server {
	rpc GetToken(token_in) returns (token_out) {}
	rpc Login(login_in) returns (login_out) {}
	rpc SelectCharacter(select_character_in) returns (select_character_out) {}
}
